tools/firmware-utils/src/ptgen.c

```diff
@@ -198,20 +332,162 @@ static int gen_ptable(uint32_t signature, int nr)
+	for (i = 0; i < nr; i++) {
+		if (!parts[i].size) {
+			if (ignore_null_sized_partition)
+				continue;
+			fprintf(stderr, "Invalid size in partition %d!\n", i);
+			return ret;
+		}
+		start = sect + sectors;
+		if (kb_align != 0)
+			start = round_to_kb(start);
+		gpte[i].start = cpu_to_le64(start);
+
+		sect = start + parts[i].size * 2;
+		if (kb_align == 0)
+			sect = round_to_cyl(sect);
+		gpte[i].end = cpu_to_le64(sect -1);
+		gpte[i].guid = guid;
+		gpte[i].guid.b[sizeof(guid_t) -1] += i + 1;
+		if (parts[i].type == 0xEF || (i + 1) == active) {
+			gpte[i].type = GUID_PARTITION_SYSTEM;
+			init_utf16("EFI System Partition", gpte[i].name, GPT_ENTRY_NAME_SIZE / sizeof(uint16_t));
+		} else {
+			gpte[i].type = GUID_PARTITION_BASIC_DATA;
+		}
+
+		if (verbose)
+			fprintf(stderr, "Partition %d: start=%lld, end=%lld, size=%lld\n",
+					i,
+					start * DISK_SECTOR_SIZE, sect * DISK_SECTOR_SIZE,
+					(sect - start) * DISK_SECTOR_SIZE);
+		printf("%lld\n", start * DISK_SECTOR_SIZE);
+		printf("%lld\n", (sect - start) * DISK_SECTOR_SIZE);
+	}
+
+	gpte[GPT_ENTRY_MAX - 1].start = cpu_to_le64(GPT_FIRST_ENTRY_SECTOR + GPT_ENTRY_SIZE * GPT_ENTRY_MAX / DISK_SECTOR_SIZE);
+	gpte[GPT_ENTRY_MAX - 1].end = cpu_to_le64((kb_align ? round_to_kb(sectors) : sectors) - 1);
+	gpte[GPT_ENTRY_MAX - 1].type = GUID_PARTITION_BIOS_BOOT;
+	gpte[GPT_ENTRY_MAX - 1].guid = guid;
+	gpte[GPT_ENTRY_MAX - 1].guid.b[sizeof(guid_t) -1] += GPT_ENTRY_MAX;
+
+	end = sect + sectors - 1;
+
+	pte.type = 0xEE;
+	pte.start = cpu_to_le32(GPT_HEADER_SECTOR);
+	pte.length = cpu_to_le32(end);
+	to_chs(GPT_HEADER_SECTOR, pte.chs_start);
+	to_chs(end, pte.chs_end);
+
+	gpth.last_usable = cpu_to_le64(end - GPT_ENTRY_SIZE * GPT_ENTRY_MAX / DISK_SECTOR_SIZE - 1);
+	gpth.alternate = cpu_to_le64(end);
+	gpth.entry_crc32 = cpu_to_le32(gpt_crc32(gpte, GPT_ENTRY_SIZE * GPT_ENTRY_MAX));
+	gpth.crc32 = cpu_to_le32(gpt_crc32((char *)&gpth, GPT_HEADER_SIZE));
+
+	if ((fd = open(filename, O_WRONLY|O_CREAT|O_TRUNC, 0644)) < 0) {
+		fprintf(stderr, "Can't open output file '%s'\n",filename);
+		return ret;
+	}
+
+	lseek(fd, MBR_DISK_SIGNATURE_OFFSET, SEEK_SET);
+	if (write(fd, &signature, sizeof(signature)) != sizeof(signature)) {
+		fputs("write failed.\n", stderr);
+		goto fail;
+	}
+
+	lseek(fd, MBR_PARTITION_ENTRY_OFFSET, SEEK_SET);
+	if (write(fd, &pte, sizeof(struct pte)) != sizeof(struct pte)) {
+		fputs("write failed.\n", stderr);
+		goto fail;
+	}
+
+	lseek(fd, MBR_BOOT_SIGNATURE_OFFSET, SEEK_SET);
+	if (write(fd, "\x55\xaa", 2) != 2) {
+		fputs("write failed.\n", stderr);
+		goto fail;
+	}
+
+	//lseek(fd, GPT_HEADER_SECTOR * DISK_SECTOR_SIZE, SEEK_SET);
+	if (write(fd, &gpth, GPT_HEADER_SIZE) != GPT_HEADER_SIZE) {
+		fputs("write failed.\n", stderr);
+		goto fail;
+	}
+
+	lseek(fd, GPT_FIRST_ENTRY_SECTOR * DISK_SECTOR_SIZE, SEEK_SET);
+	if (write(fd, &gpte, GPT_ENTRY_SIZE * GPT_ENTRY_MAX) != GPT_ENTRY_SIZE * GPT_ENTRY_MAX) {
+		fputs("write failed.\n", stderr);
+		goto fail;
+	}
+
+#if 0
```

Comment:
Why do you keep it?

Reply:
The alternate partition table is part of the gpt standard. I think that although it is omitted, the code should be retained. It may be used in the future or just as a reminder.

Reply:
ok, then please move it into separate function with descriptive name or at least name that `0` properly, like `#if WANT_ALTERNATE_PTABLE` or such